### PR TITLE
Add Assignment Rule

### PR DESCRIPTION
**Pull request by mikumikuchan**
Extracted from https://github.com/vijos/vj4/pull/331

@twd2

### CODE REVIEW

File: vj4/test/test_contest.py [hunk 1]
```diff
@@ -22,13 +22,24 @@ def time(jdoc):
 
 
 NOW = datetime.datetime.utcnow().replace(microsecond=0)
-TDOC = {'pids': [777, 778, 779], 'begin_at': NOW}
+TDOC = {'pids': [777, 778, 779],
+        'begin_at': NOW}
```
Comment (twd2) [inline vj4/test/test_contest.py]: Shouldn't be a new line.

File: vj4/test/test_contest.py [hunk 2]
```diff
@@ -118,6 +129,105 @@ def test_inject(self):
...
+    stats = contest._assignment_stat(ASSDOC, [SUBMIT_777_AC])
+    self.assertEqual(stats['score'], 22)
+    self.assertEqual(stats['penalty_score'], 22)
+    self.assertEqual(stats['time'], 2)
+    self.assertEqual(stats['detail'], [{**SUBMIT_777_AC, 'penalty_score': 22, 'time': 2}])
+
+  def test_one_late_ac(self):
+    stats = contest._assignment_stat(ASSDOC, [SUBMIT_777_AC_LATE])
+    self.assertEqual(stats['score'], 17)
+    self.assertEqual(stats['penalty_score'], 17*0.8)
```
Comment (twd2) [inline vj4/test/test_contest.py]: `17 * 0.8`

File: vj4/test/test_contest.py [hunk 3]
```diff
@@ -118,6 +129,105 @@ def test_inject(self):
...
+    self.assertEqual(stats['score'], 33+44)
+    self.assertEqual(stats['penalty_score'], 33+44)
+    self.assertEqual(stats['time'], 4+3)
+    self.assertCountEqual(stats['detail'],
+                          [{**SUBMIT_778_AC, 'penalty_score': 33, 'time': 4},
+                           {**SUBMIT_777_NAC, 'penalty_score': 44, 'time': 3}])
+
+  def test_multiple_2(self):
+    stats = contest._assignment_stat(ASSDOC, [SUBMIT_778_AC, SUBMIT_777_NAC_LATE])
+    self.assertEqual(stats['score'], 33+23)
```
Comment (twd2) [inline vj4/test/test_contest.py]: `33 + 23`

File: vj4/model/adaptor/contest.py [hunk 4]
```diff
@@ -49,6 +49,36 @@ def time(jdoc):
...
+    if j['pid'] in tdoc['pids'] and not (j['pid'] in effective and effective[j['pid']]['accept']):
+      effective[j['pid']] = j
+
+  def time(jdoc):
+    real = jdoc['rid'].generation_time.replace(tzinfo=None) - tdoc['begin_at']
+    return real.total_seconds()
+
+  def penalty_score(jdoc):
+    score = jdoc['score']
+    exceed_seconds = (jdoc['rid'].generation_time.replace(tzinfo=None) - tdoc['penalty_since']).total_seconds()
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: Where does `penalty_since` come from?

**Comment by twd2** [comment]
Please fix CI ^\_^ T\_T

File: vj4/test/test_contest.py (hunk 1 above)
Comment (mikumikuchan) [reply vj4/test/test_contest.py]: Align with ASSDOC

File: vj4/model/adaptor/contest.py (hunk 4 above)
Comment (mikumikuchan) [reply vj4/model/adaptor/contest.py]: See ASSDOC for a sample. Its a homework specific attribute.

File: vj4/test/test_contest.py (hunk 1 above)
Comment (twd2) [reply vj4/test/test_contest.py]: ???

File: vj4/test/test_contest.py [hunk 5]
```diff
@@ -22,13 +22,24 @@ def time(jdoc):
 
 
 NOW = datetime.datetime.utcnow().replace(microsecond=0)
-TDOC = {'pids': [777, 778, 779], 'begin_at': NOW}
+TDOC = {'pids': [777, 778, 779],
+        'begin_at': NOW}
+ASSDOC = {'pids': [777, 778, 779],
+          'begin_at': NOW,
```
Comment (twd2) [inline vj4/test/test_contest.py]: Shouldn't be a new line.

Comment (mikumikuchan) [reply vj4/test/test_contest.py]: for clear purpose

**Pull request merged**